lldb/source/Target/StackID.cpp

```diff
@@ -73,16 +73,34 @@ bool lldb_private::operator!=(const StackID &lhs, const StackID &rhs) {
   return lhs_scope != rhs_scope;
 }
 
-bool StackID::IsYounger(const StackID &lhs, const StackID &rhs,
-                        Process &process) {
+// BEGIN SWIFT
+/// If at least one of the stack IDs (lhs, rhs) is a heap CFA, perform the
+/// swift-specific async frame comparison. Otherwise, returns nullopt.
+static std::optional<bool>
+IsYoungerHeapCFAs(const StackID &lhs, const StackID &rhs, Process &process) {
+  const bool lhs_cfa_on_stack = lhs.IsCFAOnStack(process);
+  const bool rhs_cfa_on_stack = rhs.IsCFAOnStack(process);
+  if (lhs_cfa_on_stack && rhs_cfa_on_stack)
+    return {};
```

Comment:
this early exit is redundant to the final `return {}`.

also, this function could return `bool`, since it currently only ever returns two values: `true` or `nullopt`. But if it can only return two values, shouldn't it be true/false?

Reply:
It is true that this function only returns two values (on this patch, subsequent patches will change this though).

However, `nullopt` here really means: "the question isYoungerHeapCFAs is not applicable". So returning "false" would give the false impression that "isYoungerHeapCFAs" is returning "it's not younger".

The call site needs a way of knowing that it needs to try other comparisons, which is what nullopt represents here.

Reply:
> subsequent patches will change this though
imo that's where this change should happen

I also think this would be best served by an enum, which can be explicit about the case meanings. I think `optional<bool>` should be avoided in any language.

Reply:
another option: instead of mixing two predicates into one, have one predicate which checks if a CFA is heap, and another that checks if one heap CFAs is younger than another heap CFA.

Reply:
>one predicate which checks if a CFA is heap, and another that checks if one heap CFAs is younger than another heap CFA.

Are you suggesting something like this?

```
if (are both heap cfas(cfa1, cfa2))
   return compare heap cfas (cfa1, cfa2)
```

The problem with that is that we still need to handle the case where _one_ of them is on the heap and the other one is not. We could do:

```
if (at least one is heap cfaa(cfa1, cfa2))
   return compare heap cfas (cfa1, cfa2)
```

But then the "compare heap cfas" function would need to repeat the check to figure out which one is on the heap.

The other problem with this approach is that there will be another third case where we "don't have an opinion", which is when we were chasing continuation pointers and decided to give up because we reached some upper limit on the number of tries.

I'll try the enumeration approach if you believe that makes sense, though IMO the `optional<bool>` is quite expressive here

Reply:
Changed in the newest version